client-libraries/rust/src/plugin_utils/config.rs

```diff
@@ -0,0 +1,936 @@
+
+        let env_config = envy::from_env::<EnvConfig>()?;
+
+        // Load plugin-specific config from the 'metdata' entry
+        let mut plugin_toml = cfg.metadata.clone();
+        merge_plugin_config_from_env::<T>(env_prefix, map_env_val, &mut plugin_toml)?;
+
+        // deserialize from merged toml values to the actual struct
+        let plugin: T = TomlValue::Table(plugin_toml.into_iter().collect()).try_into()?;
+
+        // syntheisze a uuid runid if none was given
+        let run_id = env_config
+            .modality_run_id
+            .unwrap_or_else(|| uuid::Uuid::new_v4().to_string());
+
+        let client_timeout = env_config
+            .modality_client_timeout
+            .map(Duration::from_secs_f32);
+
+        Ok(Config {
+            ingest,
+            mutation,
+            plugin,
+            client_timeout,
+            run_id,
+            time_domain: env_config.modality_time_domain,
+        })
+    }
+
+    #[deprecated = "Prefer the more explicit 'connect_and_authenticate_ingest'"]
+    pub async fn connect_and_authenticate(
+        &self,
+    ) -> Result<super::ingest::Client, Box<dyn std::error::Error + Send + Sync>> {
+        self.connect_and_authenticate_ingest().await
+    }
+
+    /// Connect to the configured Modality backend for ingest,
+    /// authenticate, and return a high-level ingest client.
+    pub async fn connect_and_authenticate_ingest(
+        &self,
+    ) -> Result<super::ingest::Client, Box<dyn std::error::Error + Send + Sync>> {
+        let protocol_parent_url = if let Some(url) = &self.ingest.protocol_parent_url {
+            url.clone()
+        } else {
+            Url::parse("modality-ingest://127.0.0.1")?
+        };
+
+        // load from MODALITY_AUTH_TOKEN or from the user profile
+        let auth_token = AuthToken::load()?;
+
+        let client = IngestClient::connect_with_timeout(
+            &protocol_parent_url,
+            self.ingest.allow_insecure_tls,
+            self.client_timeout
+                .unwrap_or_else(|| Duration::from_secs(1)),
+        )
+        .await?
+        .authenticate(auth_token.into())
+        .await?;
+
+        Ok(super::ingest::Client::new(
+            client,
+            self.ingest.timeline_attributes.clone(),
+            Some(self.run_id.clone()),
+            self.time_domain.clone(),
+        )
+        .await?)
+    }
+
+    /// Connect to the configured Modality backend for mutation,
+    /// authenticate, and return a connected MutatorHost.
+    ///
+    /// This also creates an ingest connection, which is used
+    /// internally to log mutation-related events.
+    pub async fn connect_and_authenticate_mutation(
+        &self,
+    ) -> Result<super::mutation::MutatorHost, Box<dyn std::error::Error + Send + Sync>> {
+        let ingest = self.connect_and_authenticate_ingest().await?;
+
+        // TODO handle top level mutator_attributes configuration
```

Comment:
Are we tracking this TODO somewhere?

Reply:
Yeah, it's https://github.com/auxoncorp/modality/issues/2748